pkg/apis/rollouts/validation/validation.go

```diff
@@ -0,0 +1,237 @@
+	InvalidAntiAffinityStrategyMessage = "AntiAffinity must have exactly one strategy listed"
+	// InvalidAntiAffinityStrategyMessage indicates that Anti-Affinity must have weight between 1-100
+	InvalidAntiAffinityWeightMessage = "AntiAffinity weight must be between 1-100"
+	// ScaleDownLimitLargerThanRevisionLimit the message to indicate that the rollout's revision history limit can not be smaller than the rollout's scale down limit
+	ScaleDownLimitLargerThanRevisionLimit = "This rollout's revision history limit can not be smaller than the rollout's scale down limit"
+	// InvalidTrafficRoutingMessage indicates that both canary and stable service must be set to use Traffic Routing
+	InvalidTrafficRoutingMessage = "Canary service and Stable service must to be set to use Traffic Routing"
+	// InvalidAutoPromotionSecondsMessage indicates that autoPromotion must be enabled to set autoPromotionSeconds
+	InvalidAutoPromotionSecondsMessage = "Cannot set autoPromotionSeconds while autoPromotion is not enabled"
+)
+
+func ValidateRollout(rollout *v1alpha1.Rollout) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateRolloutSpec(rollout, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateRolloutSpec checks for a valid spec otherwise returns a list of errors.
+func ValidateRolloutSpec(rollout *v1alpha1.Rollout, fldPath *field.Path) field.ErrorList {
+	spec := rollout.Spec
+	allErrs := field.ErrorList{}
+	if spec.Replicas == nil {
+		message := fmt.Sprintf(MissingFieldMessage, ".Spec.Replicas")
+		allErrs = append(allErrs, field.Required(fldPath.Child("replicas"), message))
+	} else {
+		allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(*spec.Replicas), fldPath.Child("replicas"))...)
+	}
+
+	if spec.Selector == nil {
+		message := fmt.Sprintf(MissingFieldMessage, ".Spec.Selector")
+		allErrs = append(allErrs, field.Required(fldPath.Child("selector"), message))
+	} else {
+		allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(spec.Selector, fldPath.Child("selector"))...)
+		if len(spec.Selector.MatchLabels)+len(spec.Selector.MatchExpressions) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "empty selector is invalid for deployment"))
+		}
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(spec.Selector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "invalid label selector"))
+	} else {
+		data, structConvertErr := json.Marshal(&spec.Template)
+		if structConvertErr != nil {
+			allErrs = append(allErrs, field.InternalError(fldPath.Child("template"), structConvertErr))
+		}
+		var template core.PodTemplateSpec
+		structConvertErr = json.Unmarshal(data, &template)
+		if structConvertErr != nil {
+			allErrs = append(allErrs, field.InternalError(fldPath.Child("template"), structConvertErr))
```

Comment:
If the Marshal returns an error, then the Unmarshal and ValidatePodTemplateSpecForReplicaSet will almost certainly return errors since the marshal failed. I think we should stop validating the ValidatePodTemplateSpecForReplicaSet at that point since the errors from ValidatePodTemplateSpecForReplicaSet would be misleading (i.e. saying an image is required but the yaml does have an image). This makes the error more straightforward to debug.